### ISSUE TITLE
License in repo appears to be BSD 3-clause, but setup.py states it's under the PSF license?

### ISSUE THREAD

**Issue by KM4YRI** [pywin32/pypiwin32#70]
The following three license files in the repo appear to be BSD 3-clause:

* https://github.com/pywin32/pypiwin32/blob/master/LICENSE
* https://github.com/pywin32/pypiwin32/blob/master/win32/License.txt
* https://github.com/pywin32/pypiwin32/blob/master/Pythonwin/License.txt

However the [setup.py file states](https://github.com/pywin32/pypiwin32/blob/master/setup.py#L1759) that the project is under the [PSF license](https://en.wikipedia.org/wiki/Python_Software_Foundation_License).

Is the PSF in setup.py a typo?

**Comment by xoviat** [comment]
No. The PSF license is directly from [the sourceforge page](https://sourceforge.net/projects/pywin32/). I'm not a lawyer so I only understand licensing basics though.

**Comment by xoviat** [comment]
In fact the [original project page](https://pypi.org/project/pywin32/#description) has the PSF license. So you would need to convice me that everyone who worked on this project was wrong before I change the license in setup.py.

**Comment by KM4YRI** [comment]
Is Mark Hammond or Greg Stein active on this github project?  Can I summon him via: @mhammond ? Or do you have contact info?  If not I suppose that I could dig up their emails somewhere.  

The licensing thing would definitely raise an eyebrow or two so I'd like to confirm with the original authors as to what their original intention was.

**Comment by xoviat** [comment]
Mark Hammond makes commits very occasionally (and I have quite an
interesting time finding the commits on the mercurial sourceforge system,
which no one uses anymore).

> The licensing thing would definitely raise an eyebrow or two so I'd like
to confirm with the original authors as to what their original intention
was.

If you can get Mark Hammond to say "the license should be X," then I will
change the license to X. But it's quite surprising that such a widely used
library has questions about the licensing.
​

**Comment by KM4YRI** [comment]
> If you can get Mark Hammond to say "the license should be X," then I will
change the license to X. 

I will try to get in touch with him and will report back with what I find.

> But it's quite surprising that such a widely used library has questions about the licensing.

If it were a home project I probably wouldn't care, as either PSF or BSD 3-clause are great choices, honestly.  But I'm using it at work for a pretty mundane function call or two, and we have to go through the whole third-party IP license checking rigmarole.  This package is not the only one that we've found peculiarities about.  It happens more in the packages that are calling native C code from existing libraries.  But I digress.

Anyway thanks @xoviat for the quick responses.  I'll be in touch after I get ahold of Mark.
​

**Comment by mhammond** [comment]
> Mark Hammond makes commits very occasionally (and I have quite an interesting time finding the commits on the mercurial sourceforge system, which no one uses anymore).

When I do make commits, I make them in mercurial on sourceforge, which has seen ~13.5k downloads of pywin32 this week. I intend moving that to github at some time in the future, however, the existence of this pywin32 organization and other forks inside it makes things quite complicated - none of the things here are officially pywin32, and sadly there have been no contributions back to the official pywin32 project.

Re licensing - pywin32 predates the PSF license. The intent is that it becomes PSF licensed, but I haven't gone through a formal re-licensing process.

**Comment by xoviat** [comment]
> however, the existence of this pywin32 organization and other forks inside it makes things quite complicated

I have consistently offered administrative access to this organization. You have consistently not responded to that. I ask again: would you like administrative access?

**Comment by mhammond** [comment]
> I have consistently offered administrative access to this organization. You have consistently not responded to that. I ask again: would you like administrative access?

I don't understand the intent of this organization, nor how an official pywin32 repo would fit under it given the other existing repos.

**Comment by xoviat** [comment]
You can delete all of the other repositories if you host the official pywin32 repository on github.

> I don't understand the intent of this organization

The intent of the organization to allow people to look at the pywin32 source code. No one uses mercurial on sf. No one. And no one was doing anything about it.

**Comment by KM4YRI** [comment]
> When I do make commits, I make them in mercurial on sourceforge, which has seen ~13.5k downloads of pywin32 this week. I intend moving that to github at some time in the future, however, the existence of this pywin32 organization and other forks inside it makes things quite complicated - none of the things here are officially pywin32, and sadly there have been no contributions back to the official pywin32 project.

Honestly, both the Sourceforge page and the pypi.python.org page are a complete mess.  There are dead links, licensing issues (per this PR), no public link that I can find to the mercurial repositories (I get 404 not found errors), and obviously an ambiguity between the SF page and this github organization (seeing the comments above I'm starting to understand the dynamic of what's going on but it's not really documented anywere).

I'm a nobody as far as this project concerns, but I'd highly recommend Github over Sourceforge.  I think you'll find that the quantity and quality of pull requests and bug reports that you get on Github will be better than what you've got now.  13.5k downloads on Sourceforge is just the downloads, aka leeching statistics.  That says nothing of the number of contributions made or their quality.  Python.org moved to github for a good reason, as have literally 10's of thousands of other projects.  Sourceforge is a dinosaur by comparison.  Github excels in two main areas.  Letting people see the source code immediately (without hunting for the repo links), and facilitating pull requests and forking.  Sourceforge is conclusively lacking in those areas, and this hurts open source projects that are in active development.  Those are just my opinions though, like I said I'm just a dude filing a bug report.

> Re licensing - pywin32 predates the PSF license. The intent is that it becomes PSF licensed, but I haven't gone through a formal re-licensing process.

So in that case, the license files in the repo take precedence over the license definition in the setup.py file and the pypi page.  I would propose one of two things (it's your choice and the other maintainer's choice, really).  

1. Update the license to PSF in the repo
2. Change the setup.py back to BSD 3-clause until you guys are ready to update to PSF

As-is the pypi / setup.py entry is incorrect.

I'd be happy to submit a pull request for either of the above, if that would help get things moving.

**Comment by xoviat** [comment]
> I'm a nobody as far as this project concerns, but I'd highly recommend Github over Sourceforge. I think you'll find that the quantity and quality of pull requests and bug reports that you get on Github will be better than what you've got now. 13.5k downloads on Sourceforge is just the downloads, aka leeching statistics. 

I cannot agree with this enough. 

As a side note, I have made significant effort simplifying the pywin32 source tree and allowing pywin32 to compile on MSVC 2015 and later. But I am perfectly happy to throw all of that work away if @mhammond will move the official repository to Github under this organization (hence the offer to delete the other repositories). Because honestly the difficulty of using the sourceforge site and the split efforts are not helping anyone.

**Comment by mhammond** [comment]
> I'm a nobody as far as this project concerns, but I'd highly recommend Github over Sourceforge.

Yeah, that's my intent. Once Python moved to github there was no question that it's the right thing to do. I hate sourceforge as much as the next person - mercurial too :) It's just that little time I can find for pywin32 these days is making releases.

> Update the license to PSF in the repo

IANAL, but I don't think re-licensing is just a matter of updating the license.txt files. https://en.wikipedia.org/wiki/Software_relicensing - certainly it's trivial for *my* commits to be relicensed, and I'm sure we could track down many contributors.

**Comment by xoviat** [comment]
It literally takes five minutes of your time to move the repository with the GitHub import button. This repository was created using that functionality.

**Comment by xoviat** [comment]
If you want, I can create a new "pywin32" repository under this organization (I just need the mercurial URL, which escapes me at this moment) and add you as an administrator. You can then delete the other repositories. Would that be acceptable?

**Comment by KM4YRI** [comment]
> It's just that little time I can find for pywin32 these days is making releases.

Why not just delegate some of the work out to those that have time?  Of the five open issues on the github project, non really have to do with straightening those sort of issues out.  @xoviat seems willing to help, for one.  I'm quite tied up but am always willing to make room for projects that are lacking manpower.

> IANAL, but I don't think re-licensing is just a matter of updating the license.txt files. https://en.wikipedia.org/wiki/Software_relicensing - certainly it's trivial for my commits to be relicensed, and I'm sure we could track down many contributors.

I think you're right there, relicensing is a bit of a pain for the reasons that you state.  So I think the other proposed choice of changing the setup.py back to BSD 3-clause would make the most sense.  That's an easy change and would bring it to match what the work was originally licensed under.

**Comment by xoviat** [comment]
I'm actually a major fan of continuous deployment. I'm actually able to deploy a new "win32core" (read: my personal experimentations) buggy release using "git tag && git push tags." Appveyor then builds and uploads the wheels and source to pypi.

**Comment by xoviat** [comment]
Well, for the record I sent @mhammond an invitation. As the saying goes, you can lead a horse to water, but you can't make it drink:


![image](https://user-images.githubusercontent.com/15351036/30572371-c79a0db2-9cb2-11e7-8a6b-7ae9276fab2e.png)

**Comment by mhammond** [comment]
> Why not just delegate some of the work out to those that have time? Of the five open issues on the github project, non really have to do with straightening those sort of issues out.

Right - but this isn't my repo, so isn't my place to delegate. I've no idea what the state of these repos are.

> Well, for the record I sent @mhammond an invitation. As the saying goes, you can lead a horse to water, but you can't make it drink:

To be clear, you forked pywin32 and created a pywin32 organization without contacting me before hand. As far as I know, and please correct me if I'm wrong, you have not made any contributions to pywin32 at all before that date and have not contributed a single patch "upstream" since. In those circumstances, and had I created the organization first, I probably would have declined your offer to join the organization as owner/admin until I saw a track-record of contributions, in the same way contributors are usually managed for open-source projects. While I don't want to accuse you of any mal-intent, I hope you can understand my concerns here.

I'll push a pywin32 git repo soon - it will certainly require some author remapping and I'll need to consider what to do about line-endings (currently \r\n is typically used). I'll need to reflect on what to do about this organization, but I agree getting a repo up is a good first step.

**Comment by xoviat** [comment]
> In those circumstances, and had I created the organization first, I probably would have declined your offer to join the organization as owner/admin until I saw a track-record of contributions, in the same way contributors are usually managed for open-source projects.

My response to this is that I've told you that I would be perfectly fine with you removing me from this organization.

**Comment by xoviat** [comment]
If you would just accept the invitation, you can remove me from this organization if you want. It's not that complicated.

**Comment by xoviat** [comment]
Of course I wouldn't be able to do anything about it but what would be very irritating is if you accepted the offer and then did not put up an official repository.

**Comment by xoviat** [comment]
> I'll push a pywin32 git repo soon 

And by the way I haven't forgotten that this was said eight months ago and nothing happened.

**Comment by mhammond** [comment]
> If you've been offered administrative privileges to this repo, it's six of one, half-dozen the other; it wouldn't really matter who actually created it.

To be clear - there are 2 things being conflated here, the "organization" and the "repository". The issues are against the repository and I have no idea what commits are in that repo, nor what issues are open against it. The pypiwin32 repository will not turn into the official pywin32 repository, so I expect will eventually be removed.

Unless I misunderstood what you were referring to re "the five open issues"?

> I don't really see how @xoviat was doing anything malicious

I apologize if I wasn't clear or if I suggested anything else, but that's what I meant when I said "I don't want to accuse you of any mal-intent".

> And by the way I haven't forgotten that this was said eight months ago and nothing happened.

My apologies, but in that time I have been cutting pywin32 releases and been responding to issues and patches opened on sourceforge. I hope you can understand how that has been a higher priority for me than this fork. Can I count on you guys to help move all outstanding issues on sourceforge over when I publish a repo?

**Comment by xoviat** [comment]
To be clear I have offered administrative privileges on the organization. That means that if you accept the invitation you can create delete or modify any repository under the pywin32 URL.

**Comment by xoviat** [comment]
> Can I count on you guys to help move all outstanding issues on sourceforge over when I publish a repo?

If you create a repository under the pywin32 organization, I will move all of the issues over.

**Comment by xoviat** [comment]
Also the issue is that cutting a release shouldn't take but 5 minutes. Setuptools adopted this system and is able to cut hundreds of releases.

**Comment by xoviat** [comment]
> I hope you can understand how that has been a higher priority for me than this fork.

Also, we're not talking about a fork. We're talking about you creating a repository under the organization and possibly removing me because you may not trust me. I promise not to make any commits to a repository that you create in the organization but if you don't trust me that's fine.

**Comment by mhammond** [comment]
> Also the issue is that cutting a release shouldn't take but 5 minutes. Setuptools adopted this system and is able to cut hundreds of releases.

I wish - it takes hours if I clobber the world before. I'm currently a little stuck on having the docs build process work as the hg import seems to make files with \r line endings and the autoduck tool appears to get upset at that.

> Also, we're not talking about a fork.

To be clear, I was responding to the "5 issues" comment, which I assume was a reference to the pypiwin32 fork.

**Comment by xoviat** [comment]
It takes hours because you build it on your machine. You want appveyor.

**Comment by xoviat** [comment]
And for the line endings use filter branch.

**Comment by xoviat** [comment]
I am more than happy to help but I'm not going to touch mercurial or sourceforge. I can fix your problems once you have a repository.

**Comment by mhammond** [comment]
I created a repo at https://github.com/mhammond/pywin32-experimental-do-not-clone - I'll make it more formal (ie, rename it, probably move it to this org) after some feedback, but it's named that way as there's a chance I'll need to clobber it if there's a significant problem identified. I sent a mail to the existing committers - https://sourceforge.net/p/pywin32/mailman/message/36044622/

**Comment by xoviat** [comment]
Okay. Let me know if you need any help or have any questions.

**Comment by xoviat** [comment]
If you had not rewritten issue numbers in the commits the commits (#123 to
#sf123), then the imported issue numbers would have matched the issue
numbers in the commits and people would be able to click on them.

**Comment by xoviat** [comment]
@mhammond Also, would you mind posting a gist of
https://sourceforge.net/p/pywin32/admin/export?

**Comment by mhammond** [comment]
> If you had not rewritten issue numbers in the commits the commits (#123 to #sf123), then the imported issue numbers would have matched the issue numbers in the commits and people would be able to click on them.

huh - I'm surprised (but happy!) to hear they can be imported that way. I guess I can just rewrite them back.

> @mhammond Also, would you mind posting a gist of https://sourceforge.net/p/pywin32/admin/export?

Apparently gists can no longer have directories, and a full repo seems somewhat overkill, so I just uploaded the .zip directly to https://gist.github.com/mhammond/0ed20382b1e123f137bd525999e5a0e2

Thanks for your help!

**Comment by xoviat** [comment]
@mhammond Is there any update on the repository? Is it ready to go or not?

**Comment by mhammond** [comment]
I was actually waiting to hear if I should put a new one up without the issue names remapped. How has your investigation on the issue import gone?

**Comment by xoviat** [comment]
They're wasn't really an "investigation," but I looked at the Github API
and it allows editing issues. As long as no one else creates an issue
before me, I can create all 500? empty issues with API queries and then go
back and edit the original post corresponding to the sourceforge issue
content (and add comments).

The issue attachments can be transferred as well but they need storage. We
could store all of the attachments in a GitHub repository and then link to
them in the issues. Is that a good plan?

**Comment by mhammond** [comment]
To be honest I'm not really sure it's worth bothering with that - the old closed issues don't really have enough value IMO. Moving the open ones makes sense, which means they get new numbers, but I'm fine with that.

I renamed the repo so https://github.com/mhammond/pywin32 now exists. Note that I'm about to go on 4 weeks vacation and I'll look into moving the repo to this organization when I return.

Thanks!

**Comment by xoviat** [comment]
@mhammond I'll need the "read" permission on that repository in order to create issues via the GitHub API. The "read" permission allows me to assign labels, set the "assigned to" person, limit discussion on and close/open issues, but it does not allow me write access to the repository.

**Comment by xoviat** [comment]
What I've seen done before is that you can enable the "protected branches" for "master" and other branches so that you can be sure it won't be written to, and then add me as a collaborator. Then, the worst that I can do is create a new branch.

**Comment by xoviat** [comment]
@mhammond I appreciate your invitation, but as I said:

>Edit: or maybe it's better to wait as the issues would probably be deleted anyway when the repository is moved.

**Comment by xoviat** [comment]
Or actually, I can just delete this organization and we can use your personal repository if you want?

**Comment by mhammond** [comment]
> Or actually, I can just delete this organization and we can use your personal repository if you want?

Yeah, I wanted to chat about this when I get back from my vacation - an "organization" doesn't make much sense to me for pywin32 unless there's going to be many repos - and I don't see why that would be the case. Ultimately it almost makes more sense for it to be under the "python" org - but TBH I'm not familiar enough with github's vagaries to know what makes the most sense longer term.

**Comment by xoviat** [comment]
Alright then. This organization is queued for deletion (what I'll probably do it move the repositories to another organization called "pywin32-historical" and then delete all of them, but keep the organization to prevent people from taking the name).

**Comment by mhammond** [comment]
Awesome, thank you - that also means we can revisit this later and reinstate the org if desired.

Thanks!
